### ISSUE TITLE
The ImageRecordReader doesnt create a compatible DataSetIterator for lenet

### ISSUE THREAD

**Issue by learnit6** [deeplearning4j/DataVec#100]
When I run the following Datavec code, the resulting iterator and dataset will not work with the lenet mnist model.  I am using a small set of color photos in three named directories.
I get 'Score at iteration 0 is NaN' for every iteration.  I am following the examples exactly.
I notice that the rows, columns int value of the resulting dataset are 0 even though features.data is populated

```
    ImageRecordReader recordReader = new ImageRecordReader(100,100,nChannels,labelMaker);  
    recordReader.initialize(trainData);
    DataSetIterator trainIter = new RecordReaderDataSetIterator(recordReader, batchSize, 1, outputNum); 
```

model.fit(trainIter);

I used the top half as shown here:  https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/dataExamples/ImagePipelineExample.java
and the model from the Lenet Mnist training example

**Comment by learnit6** [comment]
package org.deeplearning4j.examples.convolution;

import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.InputPreProcessor;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.Updater;
import org.deeplearning4j.nn.conf.inputs.InputType;
import org.deeplearning4j.nn.conf.layers.ConvolutionLayer;
import org.deeplearning4j.nn.conf.layers.DenseLayer;
import org.deeplearning4j.nn.conf.layers.OutputLayer;
import org.deeplearning4j.nn.conf.layers.SubsamplingLayer;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.lossfunctions.LossFunctions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.io.File;
import org.datavec.api.io.filters.BalancedPathFilter;
import org.datavec.api.io.labels.ParentPathLabelGenerator;
import org.datavec.api.split.FileSplit;
import org.datavec.api.split.InputSplit;
import org.datavec.image.loader.BaseImageLoader;
import org.datavec.image.recordreader.ImageRecordReader;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;
import org.deeplearning4j.datasets.datavec.RecordReaderDataSetIterator;
import org.nd4j.linalg.factory.Nd4j;

import java.util.HashMap;
import org.deeplearning4j.nn.conf.preprocessor.CnnToFeedForwardPreProcessor;

public class LenetFolderImages {
    private static final Logger log = LoggerFactory.getLogger(LenetFolderImages.class);

    public static void main(String[] args) throws Exception {
        int nChannels = 3;
        int outputNum = 3;//number of result classes
        int batchSize = 10;
        int nEpochs = 10;
        int iterations = 1;
        int seed = 123;

        log.info("Load data....");
//        Nd4j.MAX_SLICES_TO_PRINT = -1;
//        Nd4j.MAX_ELEMENTS_PER_SLICE = -1;
//        Nd4j.ENFORCE_NUMERICAL_STABILITY = true;

        String[] allowedExt = {"jpg","gif","png"};
        // Set path to the labeled images
        File parentDir = new File("F:\\smallcat\\");
        java.util.Random rnd = new java.util.Random(seed);
        FileSplit filesInDir = new FileSplit(parentDir, allowedExt, rnd);  //specify file extensions here
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();

        // Instantiating RecordReader. Specify height and width of images.
        //RecordReader recordReader = new ImageRecordReader( 28, 28, true, labels);

        BalancedPathFilter pathFilter = new BalancedPathFilter( rnd, allowedExt, labelMaker);
        InputSplit[] filesInDirSplit = filesInDir.sample(pathFilter, 80, 20);
        InputSplit trainData = filesInDirSplit[0];
        InputSplit testData = filesInDirSplit[1];

        ImageRecordReader recordReader = new ImageRecordReader(100,100,nChannels,labelMaker);  //change channel
        recordReader.initialize(trainData);
        DataSetIterator trainIter = new RecordReaderDataSetIterator(recordReader, batchSize, 1, outputNum); //This Maybe


        log.info("Build model....");
        MultiLayerConfiguration.Builder builder = new NeuralNetConfiguration.Builder()
                .seed(seed)
                .iterations(iterations)
                .regularization(true).l2(0.0005)
                .learningRate(0.01)//.biasLearningRate(0.02)
                //.learningRateDecayPolicy(LearningRatePolicy.Inverse).lrPolicyDecayRate(0.001).lrPolicyPower(0.75)
                .weightInit(WeightInit.XAVIER)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(Updater.NESTEROVS).momentum(0.9)
                .list()
                .layer(0, new ConvolutionLayer.Builder(5, 5)
                        //nIn and nOut specify depth. nIn here is the nChannels and nOut is the number of filters to be applied
                        .nIn(nChannels)
                        .stride(1, 1)
                        .nOut(20)
                        .activation("identity")
                        .build())
                .layer(1, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(2,2)
                        .stride(2,2)
                        .build())
                .layer(2, new ConvolutionLayer.Builder(5, 5)
                        //Note that nIn need not be specified in later layers
                        .stride(1, 1)
                        .nOut(50)
                        .activation("identity")
                        .build())
                .layer(3, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(2,2)
                        .stride(2,2)
                        .build())
                .layer(4, new DenseLayer.Builder().activation("relu")
                        .nOut(500).build())
                .layer(5, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .nOut(outputNum)
                        .activation("softmax")
                        .build())
                .setInputType(InputType.convolutional(100,100,nChannels)) //See note below was conf flat (28,28,1)
                .backprop(true).pretrain(false);


        MultiLayerConfiguration conf = builder.build();
        MultiLayerNetwork model = new MultiLayerNetwork(conf);
        model.init();

        log.info("Train model....");
        model.setListeners(new ScoreIterationListener(1));

        for( int i=0; i<nEpochs; i++ ) {
            model.fit(trainIter);
            log.info("*** Completed epoch {} ***", i);
        }
    }
}

**Comment by saudet** [comment]
It doesn't look like you're normalizing your data. You'll need to make sure you do:
https://deeplearning4j.org/troubleshootingneuralnets

**Comment by learnit6** [comment]
I am close.  Iteration 1 is now a number but the rest are NaN. I am not sure why.
Is there an example I can follow to normalize the entire iterator?        
DataNormalization normalizer = new NormalizerStandardize();
        while (trainIter.hasNext()) {
            DataSet ds1 = trainIter.next();
            normalizer.fit(ds1);           //Collect the statistics (mean/stdev) from the training data. This does not modify the input data
            normalizer.transform(ds1);     //Apply normalization to the training data
            //normalizer.transform(testData);dont forget test data
            model.fit(ds1);
        }

**Comment by saudet** [comment]
https://github.com/deeplearning4j/ComputerVision-examples/blob/master/src/main/java/org/deeplearning4j/examples/cv/animalRecognition/AnimalsClassification.java

**Comment by agibsonccc** [comment]
@learnit6 you're using the normalizer completely wrong. You should be calling it on the iterator followed by iterator.reset().  Then just call transform on each minibatch.

**Comment by learnit6** [comment]
Thanks @saudet. The ImagePreProcessingScaler works great on the iterator level.  The issue must have also been in the model conf because I was still getting the NaN on iterations >1 until I swapped to the model configs in the example you suggested.  I can get it from here.  Thanks
